### PR TITLE
Expose SearchResult as Protobuf through FFI

### LINKED ISSUE #1393: SearchResult should be better exposed over the FFI
This is a couple closely related issues:

1. SearchResults are passed over the FFI as JSON, and we should pass the info as protobufs
2. The data passed over the FFI isn't the minimally useful set of data.

We might want to add some protobuf functions like this:

```
// Protobuf allows nesting these, but prost behaves weirdly if we do.
enum SearchResultReason {
    // Never used in practice. Maybe remove this from here and from the rust enum?
    KEYWORD = 0,
    ORIGIN = 1,
    URL = 2,
    PREVIOUS_USE = 3,
    BOOKMARK = 4,
    // If we get real tag support, just add `optional string tags` to SearchResult below, but
    // for now expose that it was because of tags.
    TAG = 5,
}

message SearchResult {
    required string url = 1;
    required string title = 2;
    required int64 frecency = 3;
    repeated SearchResultReason reasons = 4 [packed = true];
}

message SearchResultList {
    repeated SearchResult results = 1;
}
```

And clean up SearchResult to roughly match. I also think we should expose Reason over the FFI at the same time. This needs a little more thought though.

┆Issue is synchronized with this [Jira Story](https://jira.mozilla.com/browse/SYNC-108)

### PR DESCRIPTION
**Pull request by lyuyuan**
Fix #1393 

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
  - `swiftformat --swiftversion 4 megazords components/*/ios && swiftlint` runs without emitting any warnings or producing changes
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by eoger** [comment]
@lyuyuan is this ready for review?

**Comment by lyuyuan** [comment]
@eoger Hi! Yes, I have requested you and Thom to take a look whenever you get a chance. Thanks

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt [hunk 1]
```diff
@@ -966,31 +971,24 @@ fun stringOrNull(jsonObject: JSONObject, key: String): String? {
...
-        }
-
-        fun fromJSONArray(jsonArrayText: String): List<SearchResult> {
-            val result: MutableList<SearchResult> = mutableListOf()
-            val array = JSONArray(jsonArrayText)
-            for (index in 0 until array.length()) {
-                result.add(fromJSON(array.getJSONObject(index)))
+        fun fromMessage(msg: MsgTypes.SearchResultList): List<SearchResult> {
+            val results: MutableList<SearchResult> = mutableListOf()
+            for (index in 0 until msg.getResultsCount()) {
```
Comment (eoger) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: I think there's an array that you can get, `.results` or `.result`, here's an example of how we do collections in general:
https://github.com/mozilla/application-services/blob/2649158b61c4e6d7f87bd2e3f16553393094674b/components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/Device.kt#L77-L96

File: components/places/src/api/matcher.rs [hunk 2]
```diff
@@ -334,6 +335,28 @@ impl SearchResult {
...
+    fn from(res: SearchResult) -> Self {
+        Self {
+            url: res.url.into_string(),
+            title: res.title,
+            frecency: res.frecency,
+            reasons: res
+                .reasons
+                .iter()
+                .map(|r| match r {
+                    MatchReason::Keyword => 1,
```
Comment (eoger) [inline components/places/src/api/matcher.rs]: Can we use the enum cases instead? Example here:

https://github.com/mozilla/application-services/blob/2649158b61c4e6d7f87bd2e3f16553393094674b/components/fxa-client/src/ffi.rs#L129

File: components/places/src/places_msg_types.proto [hunk 3]
```diff
@@ -171,4 +171,26 @@ message BookmarkNodeList {
...
+    // If we get real tag support, just add `optional string tags` to SearchResult below, but
+    // for now expose that it was because of tags.
+    TAG = 6;
+}
 
+message SearchResultMessage {
+    required string url = 1;
+    required string title = 2;
+    required int64 frecency = 3;
+    repeated SearchResultReason reasons = 4 [packed = true];
```
Comment (eoger) [inline components/places/src/places_msg_types.proto]: I don't think we need `packed` anymore?

Comment (thomcc) [reply components/places/src/places_msg_types.proto]: > For historical reasons, repeated fields of scalar numeric types aren't encoded as efficiently as they could be. New code should use the special option [packed=true] to get a more efficient encoding...

According to https://developers.google.com/protocol-buffers/docs/proto

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt (hunk 1 above)
Comment (lyuyuan) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: Done

File: components/places/src/api/matcher.rs (hunk 2 above)
Comment (lyuyuan) [reply components/places/src/api/matcher.rs]: I mimicked the example above. Hope I got the idea correctly!

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt [hunk 4]
```diff
@@ -1095,31 +1100,23 @@ fun stringOrNull(jsonObject: JSONObject, key: String): String? {
 }
 
 data class SearchResult(
-    val searchString: String,
     val url: String,
     val title: String,
-    val frecency: Long,
-    val iconUrl: String? = null
+    val frecency: Long
     // Skipping `reasons` for now...
```
Comment (eoger) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: @thomcc I see we are still skipping `reasons` despite "proto-encoding" it, do we want to expose that?

**Comment by eoger** [comment]
LGTM, let's see what Thom has to say about my last comment, but this looks ready to land, thanks!

**Comment by thomcc** [comment]
This essentially looks good, but we have a bit of a code freeze due to fenix's release schedule, so we're not going to merge it until that's done. I'm leaving it in my review queue for now so that I continue to get a notification about it until then.

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt (hunk 4 above)
Comment (thomcc) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: It doesn't matter much but yes we probably should include it.

**Pull request merged**